heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/query/MetricRequest.java

```diff
@@ -0,0 +1,121 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.metricscachemgr.metricscache.query;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * immutable data bag for metric request
+ * equality: <componentName, instanceId, metricName>
+ * range: startTime ~ endTime
+ * type: aggregationGranularity
+ */
+public final class MetricRequest {
+  // The instance ids to get the stats from
+  // If nothing is specified, we will get from
+  // all the instances of the component name
+  private Map<String, Set<String>> componentNameInstanceId;
+  // What set of metrics you are interested in
+  // Example is __emit-count/default
+  private Set<String> metricNames;
+  // what timeframe data in milliseconds
+  private long startTime;
+  private long endTime;
+  private MetricGranularity aggregationGranularity;
+
+  public MetricRequest(Map<String, Set<String>> componentNameInstanceId,
+                       Set<String> metricNames,
+                       long startTime, long endTime,
+                       MetricGranularity aggregationGranularity) {
+    this.componentNameInstanceId = componentNameInstanceId;
+    this.metricNames = metricNames;
+    this.startTime = startTime;
+    this.endTime = endTime;
+    this.aggregationGranularity = aggregationGranularity;
+  }
+
+  public Map<String, Set<String>> getComponentNameInstanceId() {
+    if (componentNameInstanceId == null) {
+      return null;
+    }
+    Map<String, Set<String>> ret = new HashMap<>();
+    for (String key : componentNameInstanceId.keySet()) {
+      Set<String> value = null;
+      if (componentNameInstanceId.get(key) != null) {
+        value = new HashSet<>();
+        value.addAll(componentNameInstanceId.get(key));
+        ret.put(key, value);
+      }
+      ret.put(key, value);
+    }
+    return ret;
+  }
+
+  public Set<String> getMetricNames() {
+    if (metricNames == null) {
+      return null;
+    }
+    Set<String> ret = new HashSet<>();
```

Comment:
FYI, this could be:
```
return new HashSet<>(metricNames);
```
but why create a new set each time this is called? Is this to keep the underlying set from being updated by reference by the receiver? If so, better to initialize using google's `ImmutableSet` class (there's also `ImmutableMap`).